rewrite-maven/src/main/java/org/openrewrite/maven/AddDependencyVisitor.java

```diff
@@ -187,24 +190,31 @@ public Xml visitTag(Xml.Tag tag, ExecutionContext ctx) {
             return super.visitTag(tag, ctx);
         }
 
-        private String findVersionToUse(String groupId, String artifactId, ExecutionContext ctx) throws MavenDownloadingException {
-            if (resolvedVersion == null) {
-                if (versionComparator == null || versionComparator instanceof ExactVersion) {
-                    resolvedVersion = version;
-                } else {
-                    MavenMetadata mavenMetadata = metadataFailures == null ?
-                            downloadMetadata(groupId, artifactId, ctx) :
-                            metadataFailures.insertRows(ctx, () -> downloadMetadata(groupId, artifactId, ctx));
-                    LatestRelease latest = new LatestRelease(versionPattern);
-                    resolvedVersion = mavenMetadata.getVersioning().getVersions().stream()
-                            .filter(v -> versionComparator.isValid(null, v))
-                            .filter(v -> !Boolean.TRUE.equals(releasesOnly) || latest.isValid(null, v))
-                            .max((v1, v2) -> versionComparator.compare(null, v1, v2))
-                            .orElse(version);
-                }
-            }
+    }
 
-            return resolvedVersion;
+    interface MavenMetadataSupplier {
+        MavenMetadata get() throws MavenDownloadingException;
+    }
+
+    // VisibleForTesting
+    static String findVersionToUse(@Nullable VersionComparator versionComparator, String defaultVersion, MavenMetadataSupplier mavenMetadataSupplier,
```

Comment:
To be honest, I don't like this architectural change (I also discussed this with @Laurens-W). I think it is better for both component isolation and your mental model to keep helper methods in its own class. So in this case the `findVersionToUse` should stay in the `InsertDependencyInOrder` visitor.